### PR TITLE
Add dict decode node in BE

### PR DESCRIPTION
**Pull request by laotan332**
1. add one node, which support decode int value (which encoded by global dict) to original string value.

### CODE REVIEW

File: be/src/runtime/runtime_state.cpp [hunk 1]
```diff
@@ -194,9 +193,9 @@ Status RuntimeState::init(const TUniqueId& fragment_instance_id, const TQueryOpt
     }
 
     // Register with the thread mgr
-    if (exec_env != nullptr) {
+    if (exec_env != NULL) {
```
Comment (stdpain) [inline be/src/runtime/runtime_state.cpp:196]: use nullptr instread of NULL

File: be/src/runtime/runtime_state.h [hunk 2]
```diff
@@ -131,7 +133,7 @@ class RuntimeState {
     RuntimeProfile* runtime_profile() { return &_profile; }
 
     BufferedBlockMgr2* block_mgr2() {
-        DCHECK(_block_mgr2.get() != nullptr);
+        DCHECK(_block_mgr2.get() != NULL);
```
Comment (stdpain) [inline be/src/runtime/runtime_state.h:136]: the same problem

File: be/src/exec/vectorized/dict_decode_node.cpp [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+#include "runtime/runtime_state.h"
+#include "util/runtime_profile.h"
+
+namespace starrocks::vectorized {
+
+DictDecodeNode::DictDecodeNode(ObjectPool* pool, const TPlanNode& tnode, const DescriptorTbl& descs)
+        : ExecNode(pool, tnode, descs), _decode_node(tnode.decode_node) {}
+
+Status DictDecodeNode::init(const TPlanNode& tnode, RuntimeState* state) {
+    RETURN_IF_ERROR(ExecNode::init(tnode, state));
```
Comment (stdpain) [inline be/src/exec/vectorized/dict_decode_node.cpp:16]: It was not a good idea, init decoder in init stage, I suggest move it to prepare stage

File: be/src/exec/vectorized/dict_decode_node.cpp [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+Status DictDecodeNode::prepare(RuntimeState* state) {
+    RETURN_IF_ERROR(ExecNode::prepare(state));
+
+    std::vector<int32_t> _decode_column_ids;
+    auto global_dict = state->get_global_dict_map();
+    for (auto it : _decode_node.dict_id_to_string_ids) {
+        _encode_column_cids.emplace_back(it.first);
+        _decode_column_cids.emplace_back(it.second);
+        TypeDescriptor desc;
+        desc.type = TYPE_CHAR;
```
Comment (kangkaisen) [inline be/src/exec/vectorized/dict_decode_node.cpp:34]: TYPE_VARCHAR?

File: be/src/exec/vectorized/dict_decode_node.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+    if (*eos) {
+        *chunk = nullptr;
+        return Status::OK();
+    }
+
+    Columns decode_columns(_encode_column_cids.size());
+    for (size_t i = 0; i < _encode_column_cids.size(); i++) {
+        const ColumnPtr& encode_column = (*chunk)->get_column_by_slot_id(_encode_column_cids[i]);
+        decode_columns[i] = ColumnHelper::create_column(_encode_column_original_types[i], false);
```
Comment (kangkaisen) [inline be/src/exec/vectorized/dict_decode_node.cpp:70]: should consider nullable info.

File: be/src/exec/vectorized/dict_decode_node.h [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    struct DictDecoder {
+        using FieldType = RunTimeCppType<primitive_type>;
+        using ResultColumnType = RunTimeColumnType<result_primitive_type>;
+        using ColumnType = RunTimeColumnType<primitive_type>;
+        Dict dict;
+        Status decode(vectorized::Column* in, vectorized::Column* out) {
+            DCHECK(in != nullptr);
+            DCHECK(out != nullptr);
+            DCHECK(!in->is_nullable());
+            auto res_column = down_cast<ResultColumnType*>(out);
```
Comment (kangkaisen) [inline be/src/exec/vectorized/dict_decode_node.h:36]: should handle nullable

File: be/src/exec/vectorized/dict_decode_node.h [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                if (column->null_column_data()[i] == 0) {
+                    res_column->null_column_data()[i] = 0;
+                    FieldType key = data_column->get_data()[i];
+                    auto iter = dict.find(key);
+                    if (iter == dict.end()) {
+                        return Status::InternalError("Dict Decode failed, Dict can't take cover all key.");
+                    }
+                    res_data_column->append(iter->second);
+                } else {
+                    res_column->null_column_data()[i] = 1;
```
Comment (kangkaisen) [inline be/src/exec/vectorized/dict_decode_node.h:65]: Need call set_has_null method.

File: be/src/exec/vectorized/dict_decode_node.h [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    DictDecodeNode(ObjectPool* pool, const TPlanNode& tnode, const DescriptorTbl& descs);
+    ~DictDecodeNode() override {}
+
+    Status init(const TPlanNode& tnode, RuntimeState* state = nullptr) override;
+    Status prepare(RuntimeState* state) override;
+
+    Status open(RuntimeState* state) override;
+
+    Status get_next(RuntimeState* state, ChunkPtr* chunk, bool* eos) override;
+    Status collect_query_statistics(QueryStatistics* statistics) override;
```
Comment (kangkaisen) [inline be/src/exec/vectorized/dict_decode_node.h:22]: needn't override it

**Review by kangkaisen — DISMISSED** [review]

File: be/src/exec/vectorized/dict_decode_node.h [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+                    }
+                    res_column->append(iter->second);
+                }
+                return Status::OK();
+            }
+
+            auto column = down_cast<NullableColumn*>(in);
+            auto res_column = down_cast<NullableColumn*>(out);
+            res_column->null_column_data().resize(in->size());
+            res_column->set_has_null(true);
```
Comment (stdpain) [inline be/src/exec/vectorized/dict_decode_node.h:50]: remove set_hash_null

File: be/src/exec/vectorized/dict_decode_node.h (hunk 7 above)
Comment (stdpain) [reply be/src/exec/vectorized/dict_decode_node.h:65]: +1

Comment (stdpain) [reply be/src/exec/vectorized/dict_decode_node.h:65]: call set_null

**Review by stdpain — DISMISSED** [review]

**Review by stdpain — DISMISSED** [review]

File: be/src/runtime/global_dicts.h [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+namespace starrocks {
+namespace vectorized {
+// slice -> global dict code
+using GlobalDictMap = std::unordered_map<Slice, int, SliceHashWithSeed<PhmapSeed1>, SliceEqual>;
+// global dict code -> slice
+using RGlobalDictMap = std::unordered_map<int, Slice>;
+
+using GlobalDictMapEntity = std::pair<GlobalDictMap, RGlobalDictMap>;
+// column-id -> GlobalDictMap
+using GlobalDictMaps = std::unordered_map<uint32_t, GlobalDictMapEntity>;
```
Comment (dirtysalt) [inline be/src/runtime/global_dicts.h:15]: Could this have more meaningful name? This name looks quite similar to `GlobalDictMap`

File: be/src/exec/vectorized/dict_decode_node.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    for (size_t i = 0; i < _encode_column_cids.size(); i++) {
+        const ColumnPtr& encode_column = (*chunk)->get_column_by_slot_id(_encode_column_cids[i]);
+        decode_columns[i] = ColumnHelper::create_column(_encode_column_original_types[i], encode_column->is_nullable());
+        RETURN_IF_ERROR(_decoders[i]->decode(encode_column.get(), decode_columns[i].get()));
+    }
+
+    ChunkPtr nchunk = std::make_shared<Chunk>();
+    for (const auto& [k, v] : (*chunk)->get_slot_id_to_index_map()) {
+        if (std::find(_encode_column_cids.begin(), _encode_column_cids.end(), k) == _encode_column_cids.end()) {
+            auto& col = (*chunk)->get_column_by_slot_id(k);
```
Comment (dirtysalt) [inline be/src/exec/vectorized/dict_decode_node.cpp:77]: is that efficient enough?

File: be/src/runtime/runtime_state.cpp [hunk 12]
```diff
@@ -421,4 +420,26 @@ int64_t RuntimeState::get_load_mem_limit() const {
...
+}
+
+Status RuntimeState::init_global_dict(const GlobalDictLists& global_dict_list) {
+    for (const auto& global_dict : global_dict_list) {
+        DCHECK_EQ(global_dict.ids.size(), global_dict.strings.size());
+        vectorized::GlobalDictMap dict_map;
+        vectorized::RGlobalDictMap rdict_map;
+        int dict_sz = global_dict.ids.size();
+        for (int i = 0; i < dict_sz; ++i) {
+            auto str = _obj_pool->add(new std::string(global_dict.strings[i]));
```
Comment (dirtysalt) [inline be/src/runtime/runtime_state.cpp:434]: Could we allocate a large buffer to put all strings into it?  Looks like there are a lot of objects in object pool.

File: be/src/exec/vectorized/dict_decode_node.h [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    template <PrimitiveType primitive_type, typename Dict, PrimitiveType result_primitive_type>
+    struct DictDecoder {
+        using FieldType = RunTimeCppType<primitive_type>;
+        using ResultColumnType = RunTimeColumnType<result_primitive_type>;
+        using ColumnType = RunTimeColumnType<primitive_type>;
+        Dict dict;
+        Status decode(vectorized::Column* in, vectorized::Column* out) {
+            DCHECK(in != nullptr);
+            DCHECK(out != nullptr);
+            if (!in->is_nullable()) {
```
[the remark this answers is not in the document]
Comment (laotan332) [reply be/src/exec/vectorized/dict_decode_node.h:33]: it maybe difficulty

File: be/src/exec/vectorized/dict_decode_node.cpp (hunk 11 above)
Comment (laotan332) [reply be/src/exec/vectorized/dict_decode_node.cpp:77]: there will be few num of column，so maybe array is efficient enough？

**Review by dirtysalt — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**